Add Integration testing guide

mtalaga opened the pull request:
I've created simple doc how to integration test micronaut application using bean replacement. This was something I was looking for recently, so founding nothing in the guides I've decided to create one.

There is one thing I've spotted, don't know if it is only my env: the test script is not working. I was trying to break tests and run the script, but everytime it looked like everything passes. Are you aware of that?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/micronaut-projects/micronaut-guides?pullRequest=1009) <br/>All committers have signed the CLA.

sdelamo commented on guides/micronaut-integration-testing/micronaut-integration-testing.adoc:
```diff
@@ -0,0 +1,63 @@
+
+common:next.adoc[]
+
+Learn how to provide end-to-end tests with https://guides.micronaut.io/latest/micronaut-rest-assured-gradle-java.html[REST Assured Micronaut Tests].
```
Please, point people to [Micronaut test](https://micronaut-projects.github.io/micronaut-test/latest/guide/)

sdelamo commented on guides/micronaut-integration-testing/kotlin/src/test/kotlin/example/micronaut/HelloServiceTest.kt:
```diff
@@ -0,0 +1,45 @@
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.Test
+
+@MicronautTest // <1>
```
Use `@MicroanutTest(startApplication=false)` You don't need to start the embedded server for the test.

sdelamo commented on guides/micronaut-integration-testing/micronaut-integration-testing.adoc:
```diff
@@ -0,0 +1,63 @@
+<2> Here we are injecting beans which will be used down during the test.
+<3> Validation which is checking if we used right bean, and if the result is correct.
+
+common:runapp.adoc[]
```
since your app does not have any controller I will remove the `runapp. adoc` snippet.

sdelamo commented on guides/micronaut-integration-testing/micronaut-integration-testing.adoc:
```diff
@@ -0,0 +1,63 @@
+
+common:runapp.adoc[]
+
+common:graal-with-plugins.adoc[]
```
since your app does not have any controller I will remove the `common-graal-with-plugins. adoc` snippet.

sdelamo commented on guides/micronaut-integration-testing/groovy/src/main/groovy/example/micronaut/HelloRepository.groovy:
```diff
@@ -0,0 +1,14 @@
+@DefaultImplementation(HelloRepositoryDefault.class) // <1>
+interface HelloRepository {
+
+    String getHelloInLanguage(String language)
```
Add Constraints annotations. e.g. `@NotBlank`. I will rename the method to : 
`findHelloByLanguage` which is closer to the naming convention used in Micronaut Data repository.

sdelamo commented on guides/micronaut-integration-testing/groovy/src/main/groovy/example/micronaut/HelloRepository.groovy:
```diff
@@ -0,0 +1,14 @@
+
+    void putHelloInLanguage(String language, String hello)
+
+    void clearRepository()
```
delete this method which is only used for tests. add it to the bean replacement.

sdelamo commented on guides/micronaut-integration-testing/groovy/src/main/groovy/example/micronaut/HelloRepository.groovy:
```diff
@@ -0,0 +1,14 @@
+
+    String getHelloInLanguage(String language)
+
+    void putHelloInLanguage(String language, String hello)
```
Add Constraints annotations. e.g. `@NotBlank`. I will rename the method to : 
`save` which is closer to the naming convention used in Micronaut Data repositories.

sdelamo commented on guides/micronaut-integration-testing/groovy/src/main/groovy/example/micronaut/HelloService.groovy:
```diff
@@ -0,0 +1,15 @@
+class HelloService {
+
+    @Inject // <1>
+    HelloRepository helloRepository
```
Please, favour constructor injection vs field injection.

sdelamo reviewed (changes requested):
Thanks for the PR. I have created a couple of suggestions.

mtalaga committed:
Applied review changes

Added `@DefaultImplementation`, finished adoc with instructions

Moved classes to micronaut package, described steps in doc

Simple Integration test guide for all supported languages

mtalaga commented:
Thanks for the review, applied all your suggestions.
